service/storage_proxy.cc

```diff
@@ -5218,8 +5269,21 @@ storage_proxy::handle_read_digest(const rpc::client_info& cinfo, rpc::opt_time_p
                 }
                 auto timeout = t ? *t : db::no_timeout;
                 return p->query_result_local_digest(std::move(s), cmd, std::move(pr2.first), trace_state_ptr, timeout, da);
-            }).finally([&trace_state_ptr, src_ip] () mutable {
+            }).then_wrapped([this, &trace_state_ptr, src_ip] (future<rpc::tuple<query::result_digest, api::timestamp_type, cache_temperature>> f) mutable {
+                using final_tuple_type = rpc::tuple<query::result_digest, api::timestamp_type, cache_temperature, std::optional<replica::exception_variant>>;
                 tracing::trace(trace_state_ptr, "read_digest handling is done, sending a response to /{}", src_ip);
+                if (!f.failed()) {
+                    auto&& [result, timestamp, temp] = f.get();
+                    return make_ready_future<final_tuple_type>(std::move(result), timestamp, std::move(temp), std::nullopt);
+                } else {
+                    auto eptr = f.get_exception();
+                    if (features().typed_errors_in_read_rpc) {
+                        if (auto ex = replica::try_encode_replica_exception(eptr)) {
+                            return make_ready_future<final_tuple_type>(query::result_digest(), api::missing_timestamp, cache_temperature::invalid(), std::make_optional(std::move(ex)));
+                        }
+                    }
+                    return make_exception_future<final_tuple_type>(std::move(eptr));
+                }
```

Comment:
We have almost exactly same code three times. I wonder if we can put it in a function. Types are a little bit different, so it may be tricky and not worth it.

Reply:
Done. I think it didn't turn out too bad.